js/core/option_manager.js

```diff
@@ -0,0 +1,195 @@
+import { extend } from "./utils/extend";
+
+export class OptionManager {
+    constructor(options, optionsByReference, deprecatedOptions) {
+        this._options = options;
+        this._optionsByReference = optionsByReference;
+        this._deprecatedOptions = deprecatedOptions;
+        this._changingCallbacks = createCallBack({ syncStrategy: true });
+        this._changedCallbacks = createCallBack({ syncStrategy: true });
+        this._deprecatedCallbacks = createCallBack({ syncStrategy: true });
+        this._cachedDeprecateNames = [];
+        this.cachedGetters = {};
+        this.cachedSetters = {};
+    }
+
+    _notifyDeprecated(option) {
+        const info = this._deprecatedOptions[option];
+        if(info) {
+            this._deprecatedCallbacks.fire(option, info);
+        }
+    }
+
+    _clearField(options, name) {
+        delete options[name];
+
+        const previousFieldName = this._getParentName(name);
+        const fieldObject = previousFieldName ? this._getValue(options, previousFieldName, false) : options;
+
+        if(fieldObject) {
+            delete fieldObject[this._getFieldName(name)];
+        }
+    }
+
+    _getParentName(fullName) {
+        return fullName.substr(0, fullName.lastIndexOf('.'));
+    }
+
+    _getFieldName(fullName) {
+        const splitNames = fullName.split('.');
+        return splitNames[splitNames.length - 1];
+    }
+
+    _setField(options, fullName, value) {
+        let fieldName = "";
+        let fieldObject;
+
+        do {
+            if(fieldName) {
+                fieldName = "." + fieldName;
+            }
+
+            fieldName = this._getFieldName(fullName) + fieldName;
+            fullName = this._getParentName(fullName);
+            fieldObject = fullName ? this._getValue(options, fullName, false) : options;
+
+        } while(!fieldObject);
+
+        fieldObject[fieldName] = value;
+    }
+
+    _setValue(name, value, merge) {
+        if(!this.cachedSetters[name]) {
+            this.cachedSetters[name] = coreDataUtils.compileSetter(name);
+        }
+
+        const path = name.split(/[.[]/);
+        merge = typeUtils.isDefined(merge) ? merge : !this._optionsByReference[name];
+
+        this.cachedSetters[name](this._options, value, {
+            functionsAsIs: true,
+            merge,
+            unwrapObservables: path.length > 1 && !!this._optionsByReference[path[0]]
+        });
+    }
+
+    _setPreparedValue(name, value, merge) {
+        const previousValue = this._getValue(this._options, name, false);
+
+        if(coreDataUtils.equals(previousValue, value)) {
+            return;
+        }
+
+        this._changingCallbacks.fire(name, previousValue, value);
+
+        this._setValue(name, value, merge);
+        this._changedCallbacks.fire(name, value, previousValue);
+    }
+
+    _setRelevantNames(options, name, value) {
+        if(!name) return;
+        const alias = this._normalizeName(name);
```

Comment:
As for me `alias` has a different meaning from `normalize`. I propose to use only one word i.e. rename variable to `normalizedName` or function to `getAlias`